experimental/HasseSchmidt/src/HasseSchmidt.jl

```diff
@@ -0,0 +1,164 @@
+export hasse_derivatives
+
+### Implementation of Hasse-Schmidt derivatives as seen in 
+###
+###     Fruehbis-Krueger, Ristau, Schober: 'Embedded desingularization for arithmetic surfaces -- toward a parallel implementation'
+
+
+################################################################################
+### HASSE-SCHMIDT derivatives for single polynomials
+
+@doc raw"""
+    hasse_derivatives(f::MPolyRingElem)
+
+Return a list of Hasse-Schmidt derivatives of `f`, each with a multiindex `[a_1, ..., a_n]`, where `a_i` describes the number of times `f` was derived w.r.t. the `i`-th variable.
+
+# Examples
+```jldoctest
+julia> R, (x, y) = polynomial_ring(ZZ, ["x", "y"]);
+
+julia> f = 5*x^2 + 3*y^5;
+
+julia> hasse_derivatives(f)
+8-element Vector{Vector{Any}}:
+ [[0, 0], 5*x^2 + 3*y^5]
+ [[0, 1], 15*y^4]
+ [[0, 2], 30*y^3]
+ [[0, 3], 30*y^2]
+ [[0, 4], 15*y]
+ [[0, 5], 3]
+ [[1, 0], 10*x]
+ [[2, 0], 5]
+```
+"""
+function hasse_derivatives(f::MPolyRingElem)
+  R = parent(f)
+  n = ngens(R)
+  # define new ring with more variables: R[x1, ..., xn] -> R[x1, ..., xn, t1, ..., tn]
+  Rtemp, _ = polynomial_ring(R, "y" => 1:n, "t" => 1:n)
+  # replace f(x_i) -> f(y_i + t_i)
+  F = evaluate(f, gens(Rtemp)[1:n] + gens(Rtemp)[n+1:2n])
```

Comment:
```suggestion
  Rtemp, (y, t) = polynomial_ring(R, :y => 1:n, :t => 1:n)
  # replace f(x_i) -> f(y_i + t_i)
  F = evaluate(f, y + t)
```
This avoids allocations.

First generating the polynomial ring is cheaper with symbols instead of strings:
```
julia> n=2; R=QQ ; @time S,_ = polynomial_ring(R, "y" => 1:n, "t" => 1:n);
  0.000070 seconds (85 allocations: 3.562 KiB)

julia> n=2; R=QQ ; @time S,_ = polynomial_ring(R, :y => 1:n, :t => 1:n);
  0.000059 seconds (53 allocations: 2.328 KiB)
```

Secondly, `gens(Rtemp)` is not free either, it allocates a fresh vector and a fresh bunch of generator polynomials each time. E.g. if the coefficient ring is `QQ`, then `gens(S)` above performs $3n+1$ allocations.